[SPARK-31935][SQL][3.0] Hadoop file system config should be effective in data source options

gengliangwang opened the pull request:
<!--
Thanks for sending a pull request!  Here are some tips for you:
  1. If this is your first time, please read our contributor guidelines: https://spark.apache.org/contributing.html
  2. Ensure you have added or run the appropriate tests for your PR: https://spark.apache.org/developer-tools.html
  3. If the PR is unfinished, add '[WIP]' in your PR title, e.g., '[WIP][SPARK-XXXX] Your PR title ...'.
  4. Be sure to keep the PR description updated to reflect all changes.
  5. Please write your PR title to summarize what this PR proposes.
  6. If possible, provide a concise example to reproduce the issue for a faster review.
  7. If you want to add a new configuration, please read the guideline first for naming configurations in
     'core/src/main/scala/org/apache/spark/internal/config/ConfigEntry.scala'.
-->

### What changes were proposed in this pull request?
<!--
Please clarify what changes you are proposing. The purpose of this section is to outline the changes and how this PR fixes the issue. 
If possible, please consider writing useful notes for better and faster reviews in your PR. See the examples below.
  1. If you refactor some codes with changing classes, showing the class hierarchy will help reviewers.
  2. If you fix some SQL features, you can provide some references of other DBMSes.
  3. If there is design documentation, please add the link.
  4. If there is a discussion in the mailing list, please add the link.
-->
Mkae Hadoop file system config effective in data source options.

From `org.apache.hadoop.fs.FileSystem.java`:
```
  public static FileSystem get(URI uri, Configuration conf) throws IOException {
    String scheme = uri.getScheme();
    String authority = uri.getAuthority();

    if (scheme == null && authority == null) {     // use default FS
      return get(conf);
    }

    if (scheme != null && authority == null) {     // no authority
      URI defaultUri = getDefaultUri(conf);
      if (scheme.equals(defaultUri.getScheme())    // if scheme matches default
          && defaultUri.getAuthority() != null) {  // & default has authority
        return get(defaultUri, conf);              // return default
      }
    }
    
    String disableCacheName = String.format("fs.%s.impl.disable.cache", scheme);
    if (conf.getBoolean(disableCacheName, false)) {
      return createFileSystem(uri, conf);
    }

    return CACHE.get(uri, conf);
  }
```
Before changes, the file system configurations in data source options are not propagated in `DataSource.scala`.
After changes, we can specify authority and URI schema related configurations for scanning file systems.

This problem only exists in data source V1. In V2, we already use `sparkSession.sessionState.newHadoopConfWithOptions(options)` in `FileTable`.
### Why are the changes needed?
<!--
Please clarify why the changes are needed. For instance,
  1. If you propose a new API, clarify the use case for a new API.
  2. If you fix a bug, you can clarify why it is a bug.
-->
Allow users to specify authority and URI schema related Hadoop configurations for file source reading.

### Does this PR introduce _any_ user-facing change?
<!--
Note that it means *any* user-facing change including all aspects such as the documentation fix.
If yes, please clarify the previous behavior and the change this PR proposes - provide the console output, description and/or an example to show the behavior difference if possible.
If possible, please also clarify if this is a user-facing change compared to the released Spark versions or within the unreleased branches such as master.
If no, write 'No'.
-->
Yes, the file system related Hadoop configuration in data source option will be effective on reading.

### How was this patch tested?
<!--
If tests were added, say they were added here. Please make sure to add some test cases that check the changes thoroughly including negative and positive cases if possible.
If it was tested in a way different from regular unit tests, please clarify how you tested step by step, ideally copy and paste-able, so that other reviewers can test and check, and descendants can verify in the future.
If tests were not added, please describe why they were not added and/or why it was difficult to add.
-->
Unit test

gengliangwang committed:
[SPARK-31935][SQL] Hadoop file system config should be effective in data source options

Mkae Hadoop file system config effective in data source options.

From `org.apache.hadoop.fs.FileSystem.java`:
```
  public static FileSystem get(URI uri, Configuration conf) throws IOException {
    String scheme = uri.getScheme();
    String authority = uri.getAuthority();

    if (scheme == null && authority == null) {     // use default FS
      return get(conf);
    }

    if (scheme != null && authority == null) {     // no authority
      URI defaultUri = getDefaultUri(conf);
      if (scheme.equals(defaultUri.getScheme())    // if scheme matches default
          && defaultUri.getAuthority() != null) {  // & default has authority
        return get(defaultUri, conf);              // return default
      }
    }

    String disableCacheName = String.format("fs.%s.impl.disable.cache", scheme);
    if (conf.getBoolean(disableCacheName, false)) {
      return createFileSystem(uri, conf);
    }

    return CACHE.get(uri, conf);
  }
```
Before changes, the file system configurations in data source options are not propagated in `DataSource.scala`.
After changes, we can specify authority and URI schema related configurations for scanning file systems.

This problem only exists in data source V1. In V2, we already use `sparkSession.sessionState.newHadoopConfWithOptions(options)` in `FileTable`.

Allow users to specify authority and URI schema related Hadoop configurations for file source reading.

Yes, the file system related Hadoop configuration in data source option will be effective on reading.

Unit test

Closes #28760 from gengliangwang/ds_conf.

Authored-by: Gengliang Wang <gengliang.wang@databricks.com>
Signed-off-by: Gengliang Wang <gengliang.wang@databricks.com>

gengliangwang commented:
This PR backports https://github.com/apache/spark/pull/28760 to branch-3.0

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/28347/
Test PASSed.

SparkQA commented:
**[Test build #123723 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/123723/testReport)** for PR 28776 at commit [`f6cca6b`](https://github.com/apache/spark/commit/f6cca6b5163acef655d0c0e3d6cd4848b00314e0).

SparkQA commented:
**[Test build #123723 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/123723/testReport)** for PR 28776 at commit [`f6cca6b`](https://github.com/apache/spark/commit/f6cca6b5163acef655d0c0e3d6cd4848b00314e0).
 * This patch **fails due to an unknown error code, -9**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/123723/
Test FAILed.

gengliangwang commented:
retest this please

SparkQA commented:
**[Test build #123734 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/123734/testReport)** for PR 28776 at commit [`f6cca6b`](https://github.com/apache/spark/commit/f6cca6b5163acef655d0c0e3d6cd4848b00314e0).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/28358/
Test PASSed.

HyukjinKwon reviewed (approved):
LGTM if tests pass

SparkQA commented:
**[Test build #123734 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/123734/testReport)** for PR 28776 at commit [`f6cca6b`](https://github.com/apache/spark/commit/f6cca6b5163acef655d0c0e3d6cd4848b00314e0).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/123734/
Test FAILed.

cloud-fan commented:
retest this please

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/28378/
Test PASSed.

SparkQA commented:
**[Test build #123754 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/123754/testReport)** for PR 28776 at commit [`f6cca6b`](https://github.com/apache/spark/commit/f6cca6b5163acef655d0c0e3d6cd4848b00314e0).

dongjoon-hyun commented:
Please hold on this PR because this will break Hadoop 3.2.

dongjoon-hyun commented:
Retest this please.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/28418/
Test PASSed.

SparkQA commented:
**[Test build #123794 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/123794/testReport)** for PR 28776 at commit [`f6cca6b`](https://github.com/apache/spark/commit/f6cca6b5163acef655d0c0e3d6cd4848b00314e0).

dongjoon-hyun committed:
[SPARK-31935][SQL][TESTS][FOLLOWUP] Fix the test case for Hadoop2/3

### What changes were proposed in this pull request?

This PR updates the test case to accept Hadoop 2/3 error message correctly.

### Why are the changes needed?

SPARK-31935(https://github.com/apache/spark/pull/28760) breaks Hadoop 3.2 UT because Hadoop 2 and Hadoop 3 have different exception messages.

### Does this PR introduce _any_ user-facing change?

No.

### How was this patch tested?

Pass the Jenkins with both Hadoop 2/3 or do the following manually.

**Hadoop 2.7**
```
$ build/sbt "sql/testOnly *.FileBasedDataSourceSuite -- -z SPARK-31935"
...
[info] All tests passed.
```

**Hadoop 3.2**
```
$ build/sbt "sql/testOnly *.FileBasedDataSourceSuite -- -z SPARK-31935" -Phadoop-3.2
...
[info] All tests passed.
```

Closes #28791 from dongjoon-hyun/SPARK-31935.

Authored-by: Dongjoon Hyun <dongjoon@apache.org>
Signed-off-by: Dongjoon Hyun <dongjoon@apache.org>

dongjoon-hyun commented:
@gengliangwang . Please include my follow-PR here.
- c7d45c0e0b8c077da8ed4a902503a6102becf255

gengliangwang commented:
@dongjoon-hyun sure, I have included it in this PR. Thanks.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/28427/
Test PASSed.

SparkQA commented:
**[Test build #123802 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/123802/testReport)** for PR 28776 at commit [`5623228`](https://github.com/apache/spark/commit/5623228fd0b095141a138f753b13fd77ae422ab3).

dongjoon-hyun commented:
Thanks~

SparkQA commented:
**[Test build #123794 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/123794/testReport)** for PR 28776 at commit [`f6cca6b`](https://github.com/apache/spark/commit/f6cca6b5163acef655d0c0e3d6cd4848b00314e0).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/123794/
Test FAILed.

dongjoon-hyun commented on sql/core/src/test/scala/org/apache/spark/sql/streaming/FileStreamSourceSuite.scala:
```diff
@@ -532,6 +532,18 @@ class FileStreamSourceSuite extends FileStreamSourceTest {
+      val message = intercept[java.io.IOException] {
+        spark.readStream.option("fs.defaultFS", defaultFs).text(path)
+      }.getMessage
+      assert(message == expectMessage)
```
Oops. It seems that I missed here.

dongjoon-hyun replied:
Could you fix this place and forward-port to `master`, too?

gengliangwang replied:
ok, let me do it now

dongjoon-hyun replied:
Thanks again!

SparkQA commented:
**[Test build #123802 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/123802/testReport)** for PR 28776 at commit [`5623228`](https://github.com/apache/spark/commit/5623228fd0b095141a138f753b13fd77ae422ab3).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/123802/
Test FAILed.

gengliangwang committed:
[SPARK-31935][SQL][TESTS][FOLLOWUP] Fix the test case for Hadoop2/3

This PR updates the test case to accept Hadoop 2/3 error message correctly.

SPARK-31935(#28760) breaks Hadoop 3.2 UT because Hadoop 2 and Hadoop 3 have different exception messages.
In https://github.com/apache/spark/pull/28791, there are two test suites missed the fix

No

Unit test

Closes #28796 from gengliangwang/SPARK-31926-followup.

Authored-by: Gengliang Wang <gengliang.wang@databricks.com>
Signed-off-by: Dongjoon Hyun <dongjoon@apache.org>

SparkQA commented:
**[Test build #123821 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/123821/testReport)** for PR 28776 at commit [`da8d48d`](https://github.com/apache/spark/commit/da8d48d7984dd523f44c564ead9f7d5fb9cdd4ef).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/28446/
Test PASSed.

SparkQA commented:
**[Test build #123821 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/123821/testReport)** for PR 28776 at commit [`da8d48d`](https://github.com/apache/spark/commit/da8d48d7984dd523f44c564ead9f7d5fb9cdd4ef).
 * This patch **fails due to an unknown error code, -9**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/123821/
Test FAILed.

HyukjinKwon commented:
retest this please

SparkQA commented:
**[Test build #123839 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/123839/testReport)** for PR 28776 at commit [`da8d48d`](https://github.com/apache/spark/commit/da8d48d7984dd523f44c564ead9f7d5fb9cdd4ef).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/28464/
Test PASSed.

SparkQA commented:
**[Test build #123839 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/123839/testReport)** for PR 28776 at commit [`da8d48d`](https://github.com/apache/spark/commit/da8d48d7984dd523f44c564ead9f7d5fb9cdd4ef).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/123839/
Test FAILed.

HyukjinKwon commented:
retest this please

HyukjinKwon commented:
I reverted https://github.com/apache/spark/commit/dad163f0f4a70b84f560de4f1f0bb89354e39217 from branch-3.0.

SparkQA commented:
**[Test build #123849 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/123849/testReport)** for PR 28776 at commit [`da8d48d`](https://github.com/apache/spark/commit/da8d48d7984dd523f44c564ead9f7d5fb9cdd4ef).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/28474/
Test PASSed.

SparkQA commented:
**[Test build #123849 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/123849/testReport)** for PR 28776 at commit [`da8d48d`](https://github.com/apache/spark/commit/da8d48d7984dd523f44c564ead9f7d5fb9cdd4ef).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/123849/
Test FAILed.

cloud-fan commented:
retest this please

SparkQA commented:
**[Test build #123857 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/123857/testReport)** for PR 28776 at commit [`da8d48d`](https://github.com/apache/spark/commit/da8d48d7984dd523f44c564ead9f7d5fb9cdd4ef).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/28482/
Test PASSed.

SparkQA commented:
**[Test build #123857 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/123857/testReport)** for PR 28776 at commit [`da8d48d`](https://github.com/apache/spark/commit/da8d48d7984dd523f44c564ead9f7d5fb9cdd4ef).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

dongjoon-hyun commented:
Thank you, @gengliangwang and all.
Merged to branch-3.0.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/123857/
Test PASSed.
